internal/kuberesource/mutators.go

```diff
@@ -5,45 +5,80 @@ package kuberesource
 
 import (
 	"fmt"
+	"slices"
 	"strings"
 
 	applyappsv1 "k8s.io/client-go/applyconfigurations/apps/v1"
 	applybatchv1 "k8s.io/client-go/applyconfigurations/batch/v1"
 	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
 )
 
 const (
-	exposeServiceAnnotation   = "contrast.edgeless.systems/expose-service"
-	contrastRoleAnnotationKey = "contrast.edgeless.systems/pod-role"
+	exposeServiceAnnotation      = "contrast.edgeless.systems/expose-service"
+	contrastRoleAnnotationKey    = "contrast.edgeless.systems/pod-role"
+	skipInitializerAnnotationKey = "contrast.edgeless.systems/skip-initializer"
 )
 
 // AddInitializer adds an initializer and its shared volume to the resource.
 //
 // If the resource does not contain a PodSpec, this function does nothing.
-// This function is not idempotent.
+// This function is idempotent.
 func AddInitializer(
 	resource any,
 	initializer *applycorev1.ContainerApplyConfiguration,
-) any {
-	return MapPodSpec(resource, func(spec *applycorev1.PodSpecApplyConfiguration) *applycorev1.PodSpecApplyConfiguration {
-		// Add the initializer as an init container.
-		spec.WithInitContainers(
-			initializer,
-		)
+) (res any, err error) {
+	res = MapPodSpecWithMeta(resource, func(meta *applymetav1.ObjectMetaApplyConfiguration, spec *applycorev1.PodSpecApplyConfiguration) *applycorev1.PodSpecApplyConfiguration {
+		if meta.Annotations[skipInitializerAnnotationKey] == "true" {
+			return spec
+		}
+		if spec.RuntimeClassName == nil || !strings.HasPrefix(*spec.RuntimeClassName, "contrast-cc") {
+			return spec
+		}
+
+		// Remove already existing init containers with unique initializer name.
+		spec.InitContainers = slices.DeleteFunc(spec.InitContainers, func(c applycorev1.ContainerApplyConfiguration) bool {
+			return *c.Name == *initializer.Name
+		})
+		// Add the initializer as first init container.
+		spec.InitContainers = append([]applycorev1.ContainerApplyConfiguration{*initializer}, spec.InitContainers...)
 		if len(initializer.VolumeMounts) < 1 {
 			return spec
```

Comment:
Maybe we should return an error in this unexpected case, and add a comment that this is rather a sanity check?